packages/sdk/src/db_connection_builder.ts

```diff
@@ -0,0 +1,260 @@
+   *
+   * If credentials were supplied to connect, those passed to the callback will be equivalent to the ones used to connect.
+   *
+   * If the initial connection was anonymous, a new set of credentials will be generated by the database to identify this user.
+   *
+   * The credentials passed to the callback can be saved and used to authenticate the same user in future connections.
+   *
+   * @example
+   *
+   * ```ts
+   * spacetimeDBClient.onConnect((token, identity) => {
+   *  console.log("Connected to SpacetimeDB");
+   *  console.log("Token", token);
+   *  console.log("Identity", identity);
+   * });
+   * ```
+   */
+  onConnect(
+    callback: (
+      connection: DBConnectionImpl,
+      identity: Identity,
+      token: string
+    ) => void
+  ): DBConnectionBuilder<DBConnection> {
+    this.#emitter.on('connect', callback);
+    return this;
+  }
+
+  /**
+   * Register a callback to be invoked upon an error.
+   *
+   * @example
+   *
+   * ```ts
+   * spacetimeDBClient.onError((...args: any[]) => {
+   *  stdbLogger("warn","ERROR", args);
+   * });
+   * ```
+   */
+  onConnectError(
+    callback: (...args: any[]) => void
+  ): DBConnectionBuilder<DBConnection> {
+    this.#emitter.on('connectError', callback);
+    return this;
+  }
+
+  /**
+   * Registers a callback to run when a {@link DbConnection} whose connection initially succeeded
+   * is disconnected, either after a {@link DbConnection.disconnect} call or due to an error.
+   *
+   * If the connection ended because of an error, the error is passed to the callback.
+   *
+   * The `callback` will be installed on the `DbConnection` created by `build`
+   * before initiating the connection, ensuring there's no opportunity for the disconnect to happen
+   * before the callback is installed.
+   *
+   * Note that this does not trigger if `build` fails
+   * or in cases where {@link DbConnectionBuilder.onConnectError} would trigger.
+   * This callback only triggers if the connection closes after `build` returns successfully
+   * and {@link DbConnectionBuilder.onConnect} is invoked, i.e., after the `IdentityToken` is received.
+   *
+   * To simplify SDK implementation, at most one such callback can be registered.
+   * Calling `onDisconnect` on the same `DbConnectionBuilder` multiple times throws an error.
+   *
+   * Unlike callbacks registered via {@link DbConnection},
+   * no mechanism is provided to unregister the provided callback.
+   * This is a concession to ergonomics; there's no clean place to return a `CallbackId` from this method
+   * or from `build`.
+   *
+   * @param {function(error?: Error): void} callback - The callback to invoke upon disconnection.
+   * @throws {Error} Throws an error if called multiple times on the same `DbConnectionBuilder`.
+   */
+  onDisconnect(
+    callback: (...args: any[]) => void
```

Comment:
Args should be `conn: DbConnection, error?: Error | undefined`.

Reply:
Ack, thank you

Reply:
Done! I've implemented this.